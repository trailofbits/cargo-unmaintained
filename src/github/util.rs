use anyhow::{Context, Result, anyhow};
use std::{
    env::var,
    fs::{File, OpenOptions, create_dir_all, read_to_string},
    io::{Write, stdin},
    path::PathBuf,
    sync::{LazyLock, OnceLock},
};

#[allow(clippy::unwrap_used)]
static CONFIG_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| {
    #[cfg(not(windows))]
    {
        let base_directories = xdg::BaseDirectories::new().unwrap();
        base_directories.get_config_file("cargo-unmaintained")
    }
    #[cfg(windows)]
    {
        let local_app_data = var("LOCALAPPDATA").unwrap();
        PathBuf::from(local_app_data).join("cargo-unmaintained")
    }
});

static TOKEN_PATH: LazyLock<PathBuf> = LazyLock::new(|| CONFIG_DIRECTORY.join("token.txt"));

pub(super) static PERSONAL_TOKEN: OnceLock<String> = OnceLock::new();

pub fn load_token(f: impl FnOnce(&str) -> Result<()>) -> Result<bool> {
    let token_untrimmed = if let Ok(path) = var("GITHUB_TOKEN_PATH") {
        read_to_string(&path).with_context(|| format!("failed to read {path:?}"))?
    } else if let Ok(token) = var("GITHUB_TOKEN") {
        // smoelius: Suppress warning if `CI` is set, i.e., if running on GitHub.
        if var("CI").is_err() {
            #[cfg(__warnings)]
            crate::warn!(
                "found a token in `GITHUB_TOKEN`; consider using the more secure method of \
                 setting `GITHUB_TOKEN_PATH` to the path of a file containing the token",
            );
        }
        token
    } else if TOKEN_PATH.try_exists().with_context(|| {
        format!(
            "failed to determine whether `{}` exists",
            TOKEN_PATH.display()
        )
    })? {
        read_to_string(&*TOKEN_PATH).with_context(|| format!("failed to read {TOKEN_PATH:?}"))?
    } else {
        #[cfg(__warnings)]
        crate::warn!(
            "`GITHUB_TOKEN_PATH` and `GITHUB_TOKEN` are not set and no file was found at {}; \
             archival statuses will not be checked",
            TOKEN_PATH.display()
        );
        return Ok(false);
    };
    let token = token_untrimmed.trim_end().to_owned();
    PERSONAL_TOKEN
        .set(token.clone())
        .map_err(|_| anyhow!("`load_token` was already called"))?;
    f(&token)?;
    Ok(true)
}

pub(crate) fn save_token() -> Result<()> {
    println!("Please paste a personal access token below. The token needs no scopes.");

    let mut buf = String::new();

    {
        let n = stdin()
            .read_line(&mut buf)
            .with_context(|| "failed to read stdin")?;
        assert_eq!(buf.len(), n);
    }

    create_dir_all(&*CONFIG_DIRECTORY).with_context(|| "failed to create config directory")?;

    let mut file = OpenOptions::new()
        .create(true)
        .truncate(true)
        .write(true)
        .open(&*TOKEN_PATH)
        .with_context(|| format!("failed to open `{}`", TOKEN_PATH.display()))?;
    set_permissions(&file, 0o600)?;
    file.write_all(buf.as_bytes())
        .with_context(|| format!("failed to write `{}`", TOKEN_PATH.display()))?;

    println!(
        "Personal access token written to `{}`",
        TOKEN_PATH.display()
    );

    Ok(())
}

type CargoResult<T> = Result<T>;

// smoelius: The below definitions of `set_permissions` were copied from:
// https://github.com/rust-lang/cargo/blob/1e6828485eea0f550ed7be46ef96107b46aeb162/src/cargo/util/config.rs#L1010-L1024
#[cfg(unix)]
#[cfg_attr(dylint_lib = "try_io_result", allow(try_io_result))]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    use std::os::unix::fs::PermissionsExt;

    let mut perms = file.metadata()?.permissions();
    perms.set_mode(mode);
    file.set_permissions(perms)?;
    Ok(())
}

#[cfg(not(unix))]
#[allow(unused, clippy::unnecessary_wraps)]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    Ok(())
}
