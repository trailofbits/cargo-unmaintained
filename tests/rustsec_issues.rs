use snapbox::{Data, assert_data_eq};
use std::{
    env::{remove_var, var},
    fs::write,
    path::PathBuf,
    process::Command,
};

mod util;
use util::{Tee, tee};

#[ctor::ctor]
fn initialize() {
    remove_var("CARGO_TERM_COLOR");
}

#[test]
fn rustsec_issues() {
    const PATH_STDOUT: &str = "tests/rustsec_issues.stdout";

    let mut command = Command::new("cargo");
    command.args(["run", "--example=rustsec_issues"]);

    let output = tee(command, Tee::Stdout).unwrap();

    let stdout_actual = std::str::from_utf8(&output.captured).unwrap();

    if var("BLESS").is_ok() {
        write(PATH_STDOUT, stdout_actual).unwrap();
    } else {
        assert_data_eq!(
            stdout_actual,
            Data::read_from(&PathBuf::from(PATH_STDOUT), None),
        );
    }
}
