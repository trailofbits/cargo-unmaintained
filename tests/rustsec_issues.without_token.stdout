futures-intrusive...not found
ignoring `https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts`
maligned...not found
ignoring `https://github.com/rust-lang/rust/issues/39155`
wither...not found
mbox...not found
crypto-hash...not found
ignoring `https://github.com/rust-lang/cargo/pull/11795`
malloc_buf...not found
io_ext...found
cloud-storage...not found
cloud-storage-rs...found
loopdev-erikh...not found
loopdev...not found
conrod...found
yaml-rust...not found
not found (10)
    futures-intrusive  https://github.com/rustsec/advisory-db/issues/1175
    maligned           https://github.com/rustsec/advisory-db/issues/1209
    wither             https://github.com/rustsec/advisory-db/issues/1630
    mbox               https://github.com/rustsec/advisory-db/issues/1631
    crypto-hash        https://github.com/rustsec/advisory-db/issues/1633
    malloc_buf         https://github.com/rustsec/advisory-db/issues/1635
    cloud-storage      https://github.com/rustsec/advisory-db/issues/1715
    loopdev-erikh      https://github.com/rustsec/advisory-db/issues/1821
    loopdev            https://github.com/rustsec/advisory-db/issues/1821
    yaml-rust          https://github.com/rustsec/advisory-db/issues/1921
found (3)
    io_ext             https://github.com/rustsec/advisory-db/issues/1706
    cloud-storage-rs   https://github.com/rustsec/advisory-db/issues/1715
    conrod             https://github.com/rustsec/advisory-db/issues/1871
